src/git/repo.go

```diff
@@ -0,0 +1,590 @@
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/git-town/git-town/src/command"
+	"github.com/git-town/git-town/src/util"
+	"github.com/git-town/git-town/test/helpers"
+)
+
+var repoInCurrentDir *Repo
+
+// RepoInCurrentDir provides a Repo instance in the current working directory.
+func RepoInCurrentDir() *Repo {
+	if repoInCurrentDir == nil {
+		repoInCurrentDir = &Repo{
+			Dir:   ".",
+			Shell: &command.ShellInCurrentDir{},
+		}
+	}
+	return repoInCurrentDir
+}
+
+// Repo is a Git repository that exists inside a Git environment.
+type Repo struct {
+
+	// Dir contains the path of the directory that this repository is in.
+	Dir string
+
+	// Shell runs console commands in this repo.
+	command.Shell
```

Comment:
Should we really inherit from Shell vs having it as a child property? How are you envisioning the extension of the `LoggingShell` from #1390

Reply:
I want the architecture to consist of primitives that do one thing well, are orthogonal to each other, and can be composed together into higher-level structures that maps as directly as possible to the domain (aka the Go way). The current primitives I have in mind are:
- `Shell` instances run CLI commands on the command-line in different ways: in different folders, some stream their output to the output of the app, some execute using a customized test environment. These are stateless classes consisting only of pure functions.
- `CurrentBranchTracker` instances track the current branch in a Git repository. They get notified when the current branch has changed, and can tell what the currently checked out branch is. Since these instances are stateful, there must be only one instance per actual Git repository on disk.
- `git.Repo` instances know how to call Git. For example, they know that in order to get all the Git tags, one has to run `git tag` and then [each line of the output contains one tag plus some whitespace](https://github.com/git-town/git-town/blob/kg-merge-gitrepo/src/git/repo.go#L541-L543). You give them a `Shell` instance to call Git on the command line the way you want, and a `CurrentBranchTracker` instance to signal when they change branches. "Repo" is not the right name for this. Maybe call them `git.Runner` since all they do is run Git? I'll go with that from now on to distinguish this class from repos later. All their state is extracted into things like `CurrentBranchTracker` that can be shared between different `git.Runner` instances.
- [config](https://github.com/git-town/git-town/blob/kg-merge-gitrepo/src/git/config.go) instances provide cached access to the local and global Git configuration. You give them a `Shell` instance to read/write Git configuration the way you want. Since these instances are stateful, there must be only one `config` instance per actual Git repository on disk.

Our various code bases (production code, unit tests, Cucumber tests) compose these primitives into classes that match their respective domains. 

The tests operate on many Git repositories on disk in parallel. They have many instances of a `test.Repo` class. Such a `test.Repo` class contains:
- a `MockingShell` instance
- a `CurrentBranchTracker` instance
- a `git.Runner` instance using the `MockingShell` instance and the `CurrentBranchTracker` instance
- a `config` instance

The production code has one `prod.Repo` instance that contains:
- two `Shell` instances: a `ShellInCurrentDir` instance for "hidden" Git commands  and a `LoggingShell` instance for "visible" Git commands
- a `CurrentBranchTracker` instance
- two `git.Runner` instances, one per `Shell` instance, sharing the `CurrentBranchTracker` instance. Technically we need only one `git.Runner` instance here, and it would be nice if we used only one, but I can't think of a way have it use two `Shell` instances right now. We might have to get back to this later. Would be a nice simplification.
- a `config` instance

I hope all this makes sense, it's kind of abstract to explain. My plan is to do this as a series of smaller refactor steps, followed by a number of fine-tuning rounds. What do you think?

Reply:
I have pushed updates that implement what I have described. Would like to merge soon since the changes are accumulating and this is already a massive change. Happy to change anything down the line, just exploring the space based on some initial ideas.